### PR TITLE
spid restrictions on select

### PR DESCRIPTION
**Pull request by noodlemoodle**
In session applyTo, when there is no user, set the session spid to theme spid so we can set the spid restriction based on this for serviceprovideraware.

When trying to log in, reset the spid to null for the purpose of finding the user from the localUserDAO without restriction since the spid of the user logging in will not necessarily match that of the theme

### CODE REVIEW

**Review by jlhughes — APPROVED** [review]

File: src/foam/nanos/auth/PasswordExpiryAuthService.js [hunk 1]
```diff
@@ -41,8 +41,10 @@ foam.CLASS({
     },
     {
       name: 'login',
-      javaCode:
-        `User user = (User) ((DAO) getLocalUserDAO())
+      javaCode:`
+        // on login, clear the context spid to avoid spid restrictions on user find
+        x = x.put("spid", null);
```
Comment (lchanmann) [inline src/foam/nanos/auth/PasswordExpiryAuthService.js:46]: Might be hard to reason later why clearing spid in PasswordExpiryAuthService, maybe create a decorator for just to do that and have it as the first decorator.

File: src/foam/nanos/auth/ServiceProviderAwareDAO.js [hunk 2]
```diff
@@ -144,6 +164,12 @@ foam.CLASS({
     },
     {
       name: 'find_',
+      documentation: `
+        Allow users to remove objects for which they have permission to via "serviceprovider.read.<spid>"
```
Comment (lchanmann) [inline src/foam/nanos/auth/ServiceProviderAwareDAO.js:168]:
```suggestion
        Allow users to find object for which they have permission to via "serviceprovider.read.<spid>"
```

File: src/foam/nanos/auth/ServiceProviderAwareDAO.js [hunk 3]
```diff
@@ -153,6 +179,13 @@ foam.CLASS({
     },
     {
       name: 'remove_',
+      documentation: `
+        Allow users to remove objects for which they have permission to via "serviceprovider.read.<spid>".
+
+        Alternatively, if user has the global remove permission "serviceprovider.remove.*" for ServiceProvider,
```
Comment (lchanmann) [inline src/foam/nanos/auth/ServiceProviderAwareDAO.js:185]:
```suggestion
        Alternatively, if user has the global remove permission "serviceprovider.remove.*" for ServiceProvider or permission to remove the ServiceProvider eg. "serviceprovider.remove.<spid>",
```

File: src/foam/nanos/auth/ServiceProviderAwareDAO.js [hunk 4]
```diff
@@ -192,6 +228,27 @@ foam.CLASS({
...
+          type: 'Context'
+        }
+      ],
+      type: 'foam.mlang.predicate.Predicate',
+      documentation: `
+        Returns a predicate to restrict access to only the context spid, or return null if there is none, as is the case with login
+      `,
+      javaCode: `
+        if ( x.get("spid") != null ) {
+          ((Logger) x.get("logger")).debug(this.getClass().getSimpleName(), "select", "login", "spid restrictions set to context spid.");
```
Comment (lchanmann) [inline src/foam/nanos/auth/ServiceProviderAwareDAO.js:246]:
```suggestion
          ((Logger) x.get("logger")).debug(this.getClass().getSimpleName(), "getUnauthenticatedPredicate", "spid restrictions set to context spid.");
```

File: src/foam/nanos/auth/ServiceProviderAwareDAO.js [hunk 5]
```diff
@@ -192,6 +228,27 @@ foam.CLASS({
...
+          return MLang.EQ(getPropertyInfo(), x.get("spid"));
+        }
+        ((Logger) x.get("logger")).debug(this.getClass().getSimpleName(), "select", "login", "spid restrictions disabled.");
```
Comment (lchanmann) [inline src/foam/nanos/auth/ServiceProviderAwareDAO.js:249]:
```suggestion
        ((Logger) x.get("logger")).debug(this.getClass().getSimpleName(), "getUnauthenticatedPredicate", "spid restrictions disabled.");
```

File: src/foam/nanos/auth/PasswordExpiryAuthService.js (hunk 1 above)
Comment (noodlemoodle) [reply src/foam/nanos/auth/PasswordExpiryAuthService.js:46]: True, will do that, thanks

**Review by lchanmann — APPROVED** [review]

**Pull request merged**